File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.java
import java.util.*;
import java.io.*;
public class Main {
	static int k, w, h;
	static int board[][];
	static int dist[][][];
	static StringTokenizer st;
	
	static void initialize() {
		dist = new int[k+1][h+1][w+1];
		for(int z = 0; z<=k; z++) {
			for(int y=1; y<=h; y++) {
				for(int x=1; x<=w; x++)
					dist[z][y][x] = -1;
			}
		}
	}
	
	static class Tuple{
		int y,x, count;
		public Tuple(int y,int x, int count) {
			this.y = y;
			this.x = x;
			this.count = count;
		}
	}
	
	static boolean OOB(int y,int x) {
		return y<=0 || y> h || x<=0 ||x > w;
	}
	
	static void bfs() {
		initialize();
		
		Queue<Tuple> q = new ArrayDeque<>();
		q.add(new Tuple(1,1,k));
		dist[k][1][1] = 0;
		
		int dy[] = null;
		int dx[] = null;
		
		while(!q.isEmpty()) {
			Tuple cur = q.poll();
			
			if(cur.count != 0) {
				dy = new int[] {-2,-2,-1,-1,1,1,2,2};
				dx = new int[] {-1,1,-2,2,-2,2,-1,1};
				
				for(int dir = 0; dir < 8; dir++) {
					int ny = cur.y + dy[dir];
					int nx = cur.x + dx[dir];
					int nCount = cur.count - 1;
					if(OOB(ny,nx) || board[ny][nx] == 1)
						continue;
					
					
					if(dist[nCount][ny][nx] != -1)
						continue;
					
//					System.out.printf("%d %d\n", nCount, cur.count);
					dist[nCount][ny][nx] = dist[cur.count][cur.y][cur.x] + 1;
					q.add(new Tuple(ny,nx, nCount));
					
				}
			}
			
			
			dy = new int[] {-1,1,0,0};
			dx = new int[] {0,0,-1,1};
			for(int dir = 0; dir< 4; dir++) {
				int ny = cur.y + dy[dir];
				int nx = cur.x + dx[dir];
				int nCount = cur.count;
				if(OOB(ny,nx) || board[ny][nx] == 1)
					continue;
				
				if(dist[nCount][ny][nx] != -1)
					continue;
				
//				System.out.printf("%d %d\n", nCount, cur.count);
				dist[nCount][ny][nx] = dist[cur.count][cur.y][cur.x] + 1;
				q.add(new Tuple(ny,nx, nCount));
			}
			
		}
	}
	
	static void printBoard(int board[][]) {
		for(int y=1; y<=h; y++) {
			for(int x=1; x<=w;x++) {
				System.out.printf("%-3d", board[y][x]);
			}
			System.out.println();
		}
	}
	
	public static void main(String[] args) throws IOException{
		// TODO Auto-generated method stub
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		k = Integer.parseInt(br.readLine()); 
		st = new StringTokenizer(br.readLine());
		w = Integer.parseInt(st.nextToken());
		h = Integer.parseInt(st.nextToken());
		board = new int[h+1][w+1];
		for(int y=1; y<=h;y++) {
			st = new StringTokenizer(br.readLine());
			for(int x=1; x<=w;x++) {
				board[y][x] = Integer.parseInt(st.nextToken());
			}
		}
		bfs();
		
		int ans = Integer.MAX_VALUE;
		for(int i = k; i>=0;i--) {
			if(dist[i][h][w] != -1)
				ans = Math.min(ans, dist[i][h][w]);
		}
		
		if(ans == Integer.MAX_VALUE)
			ans = -1;
		System.out.println(ans);
		

		
	}

}

File: 백준/Gold/1600. 말이 되고픈 원숭이/말이 되고픈 원숭이.py
import collections
import sys

# Set a higher recursion limit for potential deep recursion (though BFS is iterative)
# sys.setrecursionlimit(10**6) 

def solve():
    """
    Reads input, performs the 3D BFS, and prints the minimum number of moves.
    """
    try:
        # Read K
        # The input in Python is typically read from stdin.
        # k = int(sys.stdin.readline().strip())
        k = int(input())
    except EOFError:
        return
    except ValueError:
        return
    
    # Read W and H
    try:
        # w, h = map(int, sys.stdin.readline().split())
        w, h = map(int, input().split())
    except EOFError:
        # Handle case where only K is given
        return
    except ValueError:
        # Handle case where W and H line is empty or invalid
        return

    # Read the board. The Java code uses 1-based indexing for the board,
    # so we'll adjust the Python code to use a list of lists where
    # board[y][x] corresponds to the coordinate (y, x). 
    # We will pad the board with a dummy row and column to maintain 1-based indexing
    # for cleaner translation, though standard Python is 0-based.
    # The actual board will be from board[1][1] to board[h][w].

    board = [[0] * (w + 1)] # Dummy row 0
    
    for _ in range(h):
        try:
            # row_data = [0] + list(map(int, sys.stdin.readline().split()))
            row_data = [0] + list(map(int, input().split()))
        except EOFError:
            # Handle incomplete board input
            return
        except ValueError:
            # Handle invalid row data
            return

        board.append(row_data)

    # Initialize the 3D distance array 'dist'
    # dist[k_remaining][y][x] stores the minimum moves to reach (y, x) 
    # with 'k_remaining' knight moves left.
    # Dimensions: (k+1) x (h+1) x (w+1)
    # Initialize with -1 (unvisited)
    dist = [[[-1] * (w + 1) for _ in range(h + 1)] for _ in range(k + 1)]

    # Queue for BFS: stores tuples of (y, x, k_remaining)
    # y and x are 1-based indices.
    q = collections.deque([(1, 1, k)])
    dist[k][1][1] = 0

    # Knight moves (dy, dx)
    knight_moves = [
        (-2, -1), (-2, 1), (-1, -2), (-1, 2),
        (1, -2), (1, 2), (2, -1), (2, 1)
    ]

    # Monkey/King moves (4 directions: up, down, left, right)
    monkey_moves = [
        (-1, 0), (1, 0), (0, -1), (0, 1)
    ]
    
    # OOB (Out of Bounds) check
    def is_oob(y, x):
        return y <= 0 or y > h or x <= 0 or x > w

    while q:
        # cur_y, cur_x, cur_k_count correspond to Java's cur.y, cur.x, cur.count
        cur_y, cur_x, cur_k_count = q.popleft()

        # --- 1. Knight Moves (if k_remaining > 0) ---
        if cur_k_count > 0:
            next_k_count = cur_k_count - 1
            for dy, dx in knight_moves:
                ny, nx = cur_y + dy, cur_x + dx

                if is_oob(ny, nx) or board[ny][nx] == 1:
                    continue

                # Check if this state (next_k_count, ny, nx) has been visited
                if dist[next_k_count][ny][nx] == -1:
                    dist[next_k_count][ny][nx] = dist[cur_k_count][cur_y][cur_x] + 1
                    q.append((ny, nx, next_k_count))

        # --- 2. Monkey/King Moves (4 directions) ---
        next_k_count = cur_k_count # The number of knight moves remains the same
        for dy, dx in monkey_moves:
            ny, nx = cur_y + dy, cur_x + dx

            if is_oob(ny, nx) or board[ny][nx] == 1:
                continue

            # Check if this state (cur_k_count, ny, nx) has been visited
            if dist[next_k_count][ny][nx] == -1:
                dist[next_k_count][ny][nx] = dist[cur_k_count][cur_y][cur_x] + 1
                q.append((ny, nx, next_k_count))

    # --- Find the minimum answer ---
    # The target is (h, w). The answer is the minimum dist[i][h][w] for all i from 0 to k.
    ans = float('inf')
    found = False
    
    for i in range(k, -1, -1):
        if dist[i][h][w] != -1:
            ans = min(ans, dist[i][h][w])
            found = True
            
    # If the target was reachable, print the minimum moves. Otherwise, print -1.
    if found:
        print(ans)
    else:
        print(-1)

if __name__ == "__main__":
    solve()